lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -329,23 +329,37 @@ LogicalResult LowerLayersPass::runOnModuleBody(FModuleOp moduleOp,
   StringRef circuitName = circuitOp.getName();
   hw::InnerSymbolNamespace ns(moduleOp);
 
-  // A cache of nodes we've created for values which are captured, but cannot
-  // have a symbol attached to them.  This is done to avoid creating the same
-  // node multiple times.
-  DenseMap<Value, NodeOp> nodeCache;
-
-  // Get or create a node within this module.  This is used when we need to
-  // create an XMRDerefOp to an expression.
-  auto getOrCreateNodeOp = [&](Value operand, ImplicitLocOpBuilder &builder,
-                               StringRef nameHint) -> NodeOp {
-    auto it = nodeCache.find(operand);
-    if (it != nodeCache.end())
+  // A cache of values to nameable ops that can be used
+  DenseMap<Value, Operation *> nameableCache;
+
+  // Get or create a node op for a value captured by a layer block.
+  auto getOrCreateNodeOp = [&](Value operand,
+                               ImplicitLocOpBuilder &builder) -> Operation * {
+    // Use the cache hit.
+    auto it = nameableCache.find(operand);
+    if (it != nameableCache.end())
       return it->getSecond();
+
+    // Create a new node.  Put it in the cache and use it.
     OpBuilder::InsertionGuard guard(builder);
     builder.setInsertionPointAfter(operand.getDefiningOp());
-    return nodeCache
-        .insert({operand,
-                 builder.create<NodeOp>(operand.getLoc(), operand, nameHint)})
+    SmallString<16> nameHint;
+    // Try to generate a "good" name hint to use for the node.
+    if (auto *definingOp = operand.getDefiningOp()) {
```

Comment:
Nope, resolved earlier comment too soon.  This check is inconsistent with the code 1 line above the `nameHint`; declaration -- if there is no defining op, `setInsertionPointAfter` looks like it will crash.

Generally I'd rather have defensive (but tight) code that also covers the natural things one would expect it to -- vs over-eagerly depending on the current guarantees made by the calling context / surroundings -- but this code is inconsistent across 3 lines.

Consider using `setInsertionPointAfterValue` (possibly) or go all-in on assuming this is never used on a block argument.

Reply:
I like the `setInsertionPointAfterValue` approach. I think I was originally leaning towards the "this is a lambda and it only exists to make some later, extremely restrictive code easier to read". From that perspective, not checking that this isn't a blockarg (or at least asserting) seemed fine. However, there's little cost to making this more sensible.

Nice suggestion.

Reply:
Changed to `setInsertionPointAfterValue` in: https://github.com/llvm/circt/pull/8466/commits/9e867a2ac2e4b0daa7062faa1753adf779b4025f